Added functions to set colors using RGB values

nangahamandine opened the pull request:
In addition to adding the  functions based on the requirements of issue #46 , I made some updates to the previous code base based on a syntax error in the main branch.
I also updated the circle.ml file to use the Joy library in order to test the changes made. 
For some reason, after every change, I still get only the black circle displayed and not the colored circle.

Please review these changes. Would love to get your feedback

nikochiko commented on examples/circle.ml:
```diff
@@ -1,12 +1,15 @@
-  (* Open a graphics window with dimensions 800x600 *)
-  set_color black;
+  init ();
 
-  draw_circle 150 150 50;
+  (* Create a black circle *)
+  let black_circle = circle 50 in
 
-  ignore (read_line ());
-  close_graph ();
-  exit 0
+  (* Create a colored circle with a yellow fill and a red stroke *)
+  let colored_circle = circle ~fill:(255, 255, 0) ~stroke:(255, 0, 0) 50 in
+
+  (* Show both circles *)
+  show [black_circle; colored_circle];
+
+  close ()
```
Can you create a new example for your changes please, rather than over-writing an old one?

nikochiko commented on lib/shape.ml:
```diff
@@ -38,20 +38,20 @@ let render_shape s =
-  match (x, y) with
-  | Some x, Some y -> Circle { c = { x; y }; radius = r }
-  | _ -> Circle { c = { x = 0; y = 0 }; radius = r }
+  let circle ?x ?y ?fill ?stroke r =
+    match (x, y) with
+    | Some x, Some y -> Circle { c = { x; y }; radius = r; fill; stroke }
+    | _ -> Circle { c = { x = 0; y = 0 }; radius = r; fill; stroke }  
```
I think there is extra indentation here than needed.

nikochiko commented on lib/shape.ml:
```diff
@@ -2,9 +2,9 @@ open Graphics
-type rectangle = { c : point; length : int; width : int }
-type circle = { c : point; radius : int }
-type ellipse = {c : point; rx : int; ry: int}
+type circle = { c : point; radius : int; fill : (int * int * int) option; stroke : (int * int * int) option }
+type rectangle = { c : point; length : int; width : int; fill : (int * int * int) option; stroke : (int * int * int) option }
+type ellipse = { c : point; rx : int; ry: int; fill : (int * int * int) option; stroke : (int * int * int) option }
```
1. Can you please make `fill` and `stroke` non optional?
We can explicitly default to `transparent` for `fill` and `black` for `stroke`.

2. Can you please create a new type for `color`? For now it can have two type constructors: `RGB of int * int * int` and `Transparent`.

Sudha247 replied:
Can we make the type `color` something like `Transparent | Fill of int * int * int | Stroke of int * int * int`? That way we store this information in the same field in a shape. If both are active, Fill can override Stroke, I suppose.

nangahamandine replied:
Okay, sure! Thank you

nikochiko replied:
> If both are active, Fill can override Stroke, I suppose.

@sudha247 I don't think this will let us create shapes with both a border and a fill.

Sudha247 replied:
Yeah, fair enough. I should have considered the case where stroke and fill can be two colors. @nangahamandine, we should switch to the previous scheme. Sorry for the noise.

nangahamandine replied:
> Yeah, fair enough. I should have considered the case where stroke and fill can be two colors. @nangahamandine, we should switch to the previous scheme. Sorry for the noise.

Oh sure! I'll redo my contribution. And thank you for the suggestion, it made me see the code from another perspective and challenged me to learn some new concepts.

nikochiko commented on lib/shape.ml:
```diff
@@ -38,20 +38,20 @@ let render_shape s =
+    | _ -> Circle { c = { x = 0; y = 0 }; radius = r; fill; stroke }  
 
-let rectangle ?x ?y length width =
+let rectangle ?x ?y ?fill ?stroke length width =
```
We can use default values with, e.g., `?(fill = Transparent)` and `?(stroke = black)`

nikochiko commented on lib/shape.mli:
```diff
@@ -2,9 +2,9 @@ type shape
 type shapes = shape list
 
```
After adding a `color` type, you would also need to add it here and change the function signatures.

nangahamandine commented:
Okay, I'll make the changes. Thank you @nikochiko

nangahamandine commented:
I've updated the code:|

- To make fill and stroke non optional and the default for fill is transparent while for stroke is black. The shapes can now use the default values 
- I have created a new type for color. It looks like `Transparent | Fill of int * int * int | Stroke of int * int * int` as @Sudha247 proposed which can store this information in the same field in a shape. 
- This shape.mli file has been updated accordignly.
- I created a new file called colored_circle to test the changes and the circle displays perfectly fine with the colors that I defined.
- I've updated the rest of the code accordingly.

nangahamandine commented:
I've updated this PR based on the requested changes by @nikochiko .
Please review

Sudha247 commented on lib/shape.ml:
```diff
@@ -24,36 +25,59 @@ let draw_line x1 y1 x2 y2 = draw_poly_line [| (x1, y1); (x2, y2) |]
   | Circle circle ->
-      draw_circle (denormalize circle.c).x (denormalize circle.c).y
-        circle.radius
+      set_color black;
```
This is done by default in the `init` function. Rather reset it back to default after the current rendering is done?

nangahamandine replied:
Okay, got it.

Sudha247 commented on lib/shape.ml:
```diff
@@ -1,10 +1,11 @@
 open Graphics
 
 type point = { x : int; y : int }
+type color = RGB of int * int * int | Transparent
```
Should we call it something like `Default` rather than `Transparent`? For stroke, this is associated with `black`.

nangahamandine replied:
I'm not sure if I clearly understand. Did you mean that instead of using `Transparent` for the fill color, I rather use `Default`?

Sudha247 replied:
Sorry, my wording wasn't precise. I meant renaming `Transparent` to `Default` so it's useful to both stroke and fill, where stroke defaults to black and fill defaults to transparent.

nangahamandine replied:
> Sorry, my wording wasn't precise. I meant renaming `Transparent` to `Default` so it's useful to both stroke and fill, where stroke defaults to black and fill defaults to transparent.

I renamed it already. Please check my most recent commit

Sudha247 commented on lib/shape.ml:
```diff
@@ -24,36 +25,59 @@ let draw_line x1 y1 x2 y2 = draw_poly_line [| (x1, y1); (x2, y2) |]
   { x = point.x + canvas_mid.x; y = point.y + canvas_mid.y }
 
-let render_shape s =
+  let render_shape s =
```
stray space char.

nangahamandine replied:
Oh! thanks

Sudha247 commented on lib/shape.ml:
```diff
@@ -24,36 +25,59 @@ let draw_line x1 y1 x2 y2 = draw_poly_line [| (x1, y1); (x2, y2) |]
-      draw_circle (denormalize circle.c).x (denormalize circle.c).y
-        circle.radius
+      set_color black;
+      if circle.fill <> Transparent then begin
+        set_color (match circle.fill with RGB (r, g, b) -> rgb r g b | Transparent -> black);
+        fill_circle (denormalize circle.c).x (denormalize circle.c).y circle.radius;
+      end;
+      if circle.stroke <> Transparent then begin
+        set_color (match circle.stroke with RGB (r, g, b) -> rgb r g b | Transparent -> black);
+        draw_circle (denormalize circle.c).x (denormalize circle.c).y circle.radius;
+      end;
```
I think this can be simplified a bit with nested match statements. The `Transparent -> black` part is redundant I think, since the previous check makes sure it's not `Transparent`.

nangahamandine replied:
Okay, I'll try that.
